### ISSUE TITLE
arccore.discriminator incorrectly routes in cases involving optional values

### ISSUE THREAD

**Issue by ChrisRus** [Encapsule/arcccore#9]
Encapsule/arccore git repo is now published as package @encapsule/arccore on npmjs.org.

v0.1.3 discriminator library models each each of its filter's input specifications as a digraph where vertices represent namespace:type pairs and have attached properties that track which filter(s) use each namespace:type. A namespace that's optional (e.g. it specifies a ____defaultValue in the input filter spec) is thus modeled as two vertices: name:jsUndefined and name:jsString (for example). The bug is the name:jsUndefined is falsely considered as a feature of a request message destined for this particular filter.

The bug manifests if this optional value appears higher in request graph model than an actually persistent feature(s). This is particularly problematic because jsUndefined indcates the absence of _any_ value. And filter and discriminator are super explicit (or they're supposed to be). So a request message sent to discriminator that happens to not specify any value for this particular optional namespace (and most do not...) will get routed to the filter that defines this troublesome optional namespace.

v0.1.4 partially fixes this issue by completely ignoring optional namespaces when it builds the model digraph. But, this is problematic for other reasons: without these namespaces in the digraph model another filter that defines a required value using the namespace name will route incorrectly. >:/

I'll leave this open until it's actually resolved.

**Comment by ChrisRus** [comment]
This is a confirmed big in all versions of @encapsule/arccore.discriminator RTL that impacts the correct dispatch of MDR messages.

The fix is understood but is non-trivial and will take ~1-week (maybe a little more) to implement and test. After that fix is in place, them @arccore.discriminator should support MDR for any set of unique valid @encapsule/arccore.filter's whose input specifications satisfy @encapsule/arccore.discriminator rules re: feature differentiation.

**IMPORTANT**

If you are using @encapsule/arccore.discriminator for MDR-pattern DO NOT USE filter specification ____defaultValue or optional namespaces (e.g. ____defaultValue, ____types [ ... "jsUndefined" ]. Anywhere in the root namespaces of your message specifications.

In other words, make suare that each of your filters supports an unambiguously required outer structure that ensures it can be discriminated from other filters in your filter set **without encountering this issue**.

If you're planning to use MDR-pattern dispatch and have questions get in touch.
